### PR TITLE
[sw/silicon_creator] Implement set minimum required BL0 security version boot service

### PR DESCRIPTION
**Pull request by jon-flatley**
Implements the Set minimum required BL0 security version boot service.

This PR is based off https://github.com/lowRISC/opentitan/pull/19348

Resolves https://github.com/lowRISC/opentitan/issues/19156

### CODE REVIEW

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 1]
```diff
@@ -189,16 +196,74 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+static rom_error_t boot_svc_min_sec_ver_handler(boot_svc_msg_t *boot_svc_msg,
+                                                boot_data_t *boot_data) {
+  boot_data->min_security_version_bl0 =
+      boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver;
+
+  // Write boot data, updating relevant fields and recomputing the digest.
+  HARDENED_RETURN_IF_ERROR(boot_data_write(boot_data));
+  // Read the boot data back to ensure the correct policy is used on this boot.
+  HARDENED_RETURN_IF_ERROR(boot_data_read(lc_state, boot_data));
+  HARDENED_RETURN_IF_ERROR(boot_data_check(boot_data));
```
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:253]:
```suggestion
```

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 2]
```diff
@@ -189,16 +196,74 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+
+  boot_svc_min_bl0_sec_ver_res_init(kErrorOk,
+                                    &boot_svc_msg->min_bl0_sec_ver_res);
```
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:256]: Can we also include the new minimum security version that we read from flash?

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 3]
```diff
@@ -189,16 +196,74 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+
+  return kErrorOk;
```
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:258]:
```suggestion
  HARDENED_CHECK_EQ(boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver, boot_data->min_security_version_bl0);
  return kErrorOk;
```

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 4]
```diff
@@ -189,16 +196,74 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+                                     &boot_svc_msg->primary_bl0_slot_res);
+
+  return kErrorOk;
+}
+
+OT_WARN_UNUSED_RESULT
+static rom_error_t boot_svc_min_sec_ver_handler(boot_svc_msg_t *boot_svc_msg,
+                                                boot_data_t *boot_data) {
+  boot_data->min_security_version_bl0 =
+      boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver;
```
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:247]: We must use a ratcheting increment here:
```suggestion
  const uint32_t current_min_sec_ver =  boot_data->min_security_version_bl0;
  const uint32_t requested_min_sec_ver = boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver;
  const uint32_t diff = requested_min_sec_ver - current_min_sec_version;
  
  if (launder32(requested_min_sec_ver) >= current_sec_ver) {
    HARDENED_CHECK_GE(requested_min_sec_ver, current_sec_ver);
    HARDENED_CHECK_LE(diff, requested_min_sec_ver);
    ...
    // Proceed with the update.
    ...
  } else {
    return kErrorOk;
  }
```

**Review by alphan — APPROVED** [review]
Thanks @jon-flatley LGTM mod the missing ratcheting increment.

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 5]
```diff
@@ -189,16 +193,92 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+
+    boot_svc_min_bl0_sec_ver_res_init(boot_data->min_security_version_bl0,
+                                      kErrorOk,
+                                      &boot_svc_msg->min_bl0_sec_ver_res);
+
+    HARDENED_CHECK_EQ(requested_min_sec_ver,
+                      boot_data->min_security_version_bl0);
+  } else {
+    boot_svc_min_bl0_sec_ver_res_init(current_min_sec_ver, kErrorOk,
+                                      &boot_svc_msg->min_bl0_sec_ver_res);
```
Comment (jon-flatley) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:271]: @alphan Is this what you had in mind? What should we put in the response message? Should this still be `kErrorOk`?

Comment (alphan) [reply sw/device/silicon_creator/rom_ext/rom_ext.c:271]: Yes, this LGTM. Thanks @jon-flatley!

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 6]
```diff
@@ -189,16 +193,92 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
...
+  const uint32_t current_min_sec_ver = boot_data->min_security_version_bl0;
+  const uint32_t requested_min_sec_ver =
+      boot_svc_msg->next_boot_bl0_slot_req.next_bl0_slot;
+  const uint32_t diff = requested_min_sec_ver - current_min_sec_ver;
+
+  // Ensure the requested minimum security version isn't lower than the current
+  // minimum security version.
+  if (launder32(requested_min_sec_ver) >= current_min_sec_ver) {
+    HARDENED_CHECK_GE(requested_min_sec_ver, current_min_sec_ver);
+    HARDENED_CHECK_LE(diff, requested_min_sec_ver);
```
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:253]: Actually, can we make this a strict inequality?

```suggestion
  if (launder32(requested_min_sec_ver) > current_min_sec_ver) {
    HARDENED_CHECK_GT(requested_min_sec_ver, current_min_sec_ver);
    HARDENED_CHECK_LT(diff, requested_min_sec_ver);
```

**Pull request merged**